columnphysics/icepack_snow.F90

```diff
@@ -966,43 +1089,56 @@ subroutine snow_dry_metamorph (nslyr,nilyr, dt, rsnw, drsnw_dry, zqsn, &
+      do k = 1, nslyr
+         zrhos(k) = smice(k) + smliq(k)
 
-      ! if snw_aging_table = 'test'
-      ! for testing Icepack without netcdf,
-      ! use a subsampled, hard-coded table
-      !   5 temperatures
-      !   5 temperature gradients
-      !   1 snow density
+         !-----------------------------------------------------------------
+         ! best-fit table indices:
+         ! Will abort if 1D data is not regularly spaced (lin_* flag must be true)
+         ! Leave option for doing a search into non regularly spaced 1D data in the future
+         ! via a binary search or similar.
+         !-----------------------------------------------------------------
 
-      do k = 1, nslyr
-          zrhos(k) = smice(k) + smliq(k)
-
-          ! best-fit table indices:
-          T_idx    = nint(abs(zTsn(k)+ Tffresh - 223.0_dbl_kind) / 5.0_dbl_kind, kind=int_kind) - iT_shift
-          Tgrd_idx = nint(zdTdz(k) / 10.0_dbl_kind, kind=int_kind) - iTgrd_shift
-          !rhos_idx = nint(zrhos(k)-50.0_dbl_kind) / 50.0_dbl_kind, kind=int_kind) - irhos_shift ! variable density
-          rhos_idx = nint((rhos-50.0_dbl_kind) / 50.0_dbl_kind, kind=int_kind) - irhos_shift    ! fixed density
-
-          ! boundary check:
-          T_idx = min(isnw_T, max(1,T_idx+1))
-          Tgrd_idx = min(isnw_Tgrd, max(1,Tgrd_idx+1))
-          rhos_idx = min(isnw_rhos, max(1,rhos_idx+1))
-
-          bst_tau   = snowage_tau  (rhos_idx,Tgrd_idx,T_idx)
-          bst_kappa = snowage_kappa(rhos_idx,Tgrd_idx,T_idx)
-          bst_drdt0 = snowage_drdt0(rhos_idx,Tgrd_idx,T_idx)
-
-          ! change in snow effective radius, using best-fit parameters
-          dr_fresh = max(c0,rsnw(k)-rsnw_fall)
-          drsnw_dry(k) = (bst_drdt0*(bst_tau/(dr_fresh+bst_tau))**(1/bst_kappa))&
-                     * (dt/3600.0_dbl_kind)
-         enddo
+         if (lin_rhos) then
+            rhos_idx = nint(   (rhos             - min_rhos) / del_rhos, kind=int_kind) + 1
+         else
+            call icepack_warnings_setabort(.true.,__FILE__,__LINE__)
+            call icepack_warnings_add(subname//'ERROR: nonlinear lookup table for rhos not supported yet')
+            return
+         endif
+
+         if (lin_Tgrd) then
+            Tgrd_idx = nint(   (zdTdz(k)         - min_Tgrd) / del_Tgrd, kind=int_kind) + 1
+         else
+            call icepack_warnings_setabort(.true.,__FILE__,__LINE__)
+            call icepack_warnings_add(subname//'ERROR: nonlinear lookup table for Tgrd not supported yet')
+            return
+         endif
+
+         if (lin_T) then
+            T_idx    = nint(abs(zTsn(k)+ Tffresh - min_T   ) / del_T   , kind=int_kind) + 1
+         else
+            call icepack_warnings_setabort(.true.,__FILE__,__LINE__)
+            call icepack_warnings_add(subname//'ERROR: nonlinear lookup table for T not supported yet')
+            return
+         endif
+
+         ! boundary check:
+         rhos_idx = min(isnw_rhos, max(1,rhos_idx))
```

Comment:
These indices are different by +1.  Is this a bug in the original code?

Reply:
I moved the +1 into the index calculation.  If you look at the previous implementation, it was part of the index check in the next code block.  I thought it was clearer to put the +1 in the computation of the index then check the index cleanly vs min/max after.